Assets/LeapMotionModules/UI/Interaction/InteractionHand.cs

```diff
@@ -0,0 +1,780 @@
+
+    #region Contact Callbacks
+
+    private Dictionary<InteractionBehaviourBase, int> _contactBehaviours = new Dictionary<InteractionBehaviourBase, int>();
+    private HashSet<InteractionBehaviourBase> _contactBehavioursLastFrame = new HashSet<InteractionBehaviourBase>();
+    private List<InteractionBehaviourBase>    _contactBehaviourRemovalCache = new List<InteractionBehaviourBase>();
+
+    internal void ContactBoneCollisionEnter(BrushBone contactBone, InteractionBehaviourBase interactionObj) {
+      int count;
+      if (_contactBehaviours.TryGetValue(interactionObj, out count)) {
+        _contactBehaviours[interactionObj] = count + 1;
+      }
+      else {
+        _contactBehaviours[interactionObj] = 1;
+      }
+    }
+
+    internal void ContactBoneCollisionExit(BrushBone contactBone, InteractionBehaviourBase interactionObj) {
+      int count = _contactBehaviours[interactionObj];
+      if (count == 1) {
+        _contactBehaviours.Remove(interactionObj);
+      }
+      else {
+        _contactBehaviours[interactionObj] = count - 1;
+      }
+    }
+
+    // TODO: Is there a need for separate methods for Triggers?
+    //internal void ContactBoneTriggerEnter(ContactBone contactBone, InteractionBehaviourBase interactionObjr) {
+    //}
+
+    //internal void ContactBoneTriggerExit(ContactBone contactBone, InteractionBehaviourBase interactionObj) {
+    //}
+
+    private void FixedUpdateContactCallbacks() {
+      foreach (var interactionObj in _contactBehavioursLastFrame) {
+        if (!_contactBehaviours.ContainsKey(interactionObj)) {
+          interactionObj.ContactEnd(_hand);
+          _contactBehaviourRemovalCache.Add(interactionObj);
+        }
+      }
+      foreach (var interactionObj in _contactBehaviourRemovalCache) {
+        _contactBehavioursLastFrame.Remove(interactionObj);
+      }
+      _contactBehaviourRemovalCache.Clear();
+      foreach (var intObjCountPair in _contactBehaviours) {
+        var interactionObj = intObjCountPair.Key;
+        if (_contactBehavioursLastFrame.Contains(interactionObj)) {
+          interactionObj.ContactStay(_hand);
+        }
+        else {
+          interactionObj.ContactBegin(_hand);
+          _contactBehavioursLastFrame.Add(interactionObj);
+        }
+      }
+    }
+
+    #endregion
+
+    #endregion
+
+    #region Grasping
+
+    private HeuristicGrabClassifier _grabClassifier;
+    private InteractionBehaviourBase _graspedObject;
+
+    private void InitGrasping() {
+      _grabClassifier = new HeuristicGrabClassifier(this);
+    }
+
+    private void FixedUpdateGrasping() {
+      _grabClassifier.FixedUpdate();
+
+      if (_graspedObject != null) {
+        _graspedObject.GraspHold(_hand);
+      }
+    }
+
+    public void Grasp(InteractionBehaviourBase interactionObj) {
+      interactionObj.GraspBegin(_hand);
+      _graspedObject = interactionObj;
+    }
+
+    public void ReleaseGrasp() {
+      if (_graspedObject == null) return; // Nothing to release.
```

Comment:
I think that it should be an error to tell a hand to release if it is not currently releasing.  Either that, or make this a Try method so that it is obvious that the method can have no effect.